src/webauthn/proto/raw_message.rs

```diff
@@ -45,14 +58,525 @@ pub struct Packed {
+        cursor.read_exact(&mut auth_policy)?;
+
+        let (parameters, public_id) = match TpmAlgId::from_u16(alg_type) {
+            TpmAlgId::RSA => {
+                let parameters = AlgParameters::RSA(RsaAlgParameters {
+                    symmetric: TpmAlgId::from_u16(cursor.read_u16::<BigEndian>()?),
+                    scheme: TpmAlgId::from_u16(cursor.read_u16::<BigEndian>()?),
+                    key_bits: cursor.read_u16::<BigEndian>()?,
+                    exponent: cursor.read_u32::<BigEndian>()?,
+                });
+
+                let unique_length = cursor.read_u16::<BigEndian>()?;
+                let mut unique = vec![0u8; unique_length as usize];
+                cursor.read_exact(&mut unique)?;
+
+                (parameters, TpmuPublicId::Rsa(unique))
+            }
+            TpmAlgId::ECC => {
+                let parameters = AlgParameters::ECC(EccAlgParameters {
+                    symmetrics: TpmAlgId::from_u16(cursor.read_u16::<BigEndian>()?),
+                    scheme: TpmAlgId::from_u16(cursor.read_u16::<BigEndian>()?),
+                    curve_id: TpmEccCurve::from_u16(cursor.read_u16::<BigEndian>()?),
+                    kdf: TpmAlgId::from_u16(cursor.read_u16::<BigEndian>()?),
+                });
+
+                let x_length = cursor.read_u16::<BigEndian>()?;
+                let mut x = vec![0u8; x_length as usize];
+                cursor.read_exact(&mut x)?;
+
+                let y_length = cursor.read_u16::<BigEndian>()?;
+                let mut y = vec![0u8; y_length as usize];
+                cursor.read_exact(&mut y)?;
+
+                (parameters, TpmuPublicId::Ecc(EccPoint { x, y }))
+            }
+            _ => (AlgParameters::None, TpmuPublicId::None),
+        };
+
+        Ok(PublicArea {
+            alg_type,
+            name_alg,
+            object_attributes,
+            auth_policy: TPM2BDigest {
+                size: auth_policy_length,
+                buffer: Some(auth_policy),
+            },
+            parameters,
+            unique: public_id,
+        })
+    }
+}
+
+#[derive(Serialize, Deserialize, Default, Clone, Debug)]
+pub enum TpmuPublicId {
+    #[default]
+    None,
+    Rsa(Vec<u8>),
+    Ecc(EccPoint),
+}
+
+#[derive(Serialize, Deserialize, Clone, Debug)]
+pub struct EccPoint {
+    pub x: Vec<u8>,
+    pub y: Vec<u8>,
+}
+
+#[derive(Default, Serialize, Deserialize, Clone, Debug)]
+#[serde(rename_all = "camelCase")]
+pub struct CertInfo {
+    pub magic: u32,
+    pub attestation_type: AttestationType,
+    pub qualified_signer: Tpm2bName,
+    pub extra_data: Tpm2bData,
+    pub clock_info: TpmsClockInfo,
+    pub firmware_version: u64,
+    pub alg: u16,
+    pub attested_name: (Vec<u8>, Vec<u8>),
+}
+
+impl CertInfo {
+    pub fn from_buf(buf: &[u8]) -> Result<CertInfo, Error> {
+        CertInfo::from_vec(buf.to_vec())
+    }
+
+    pub fn from_vec(buf: Vec<u8>) -> Result<CertInfo, Error> {
```

Comment:
```suggestion
    pub fn from_slice(buf: &[u8]) -> Result<CertInfo, Error> {
```
Since we only read, we can take a slice instead, so the caller don't have to copy the vec.